pkg/vm/engine/tae/containers/vector.go

```diff
@@ -16,411 +16,571 @@ package containers
+	if err != nil {
+		return nil
 	}
-	buf = buf[nr:]
-	if !nullable {
-		vec.roStorage = n
+
+	return cloned
+}
+
+//TODO: --- Below Functions Can be implemented in CN Vector.
+
+func (vec *vector[T]) ExtendWithOffset(src Vector, srcOff, srcLen int) {
+	//TODO: CN vector include
+
+	if srcLen <= 0 {
 		return
 	}
 
-	nullSize := types.DecodeFixed[uint32](buf[:4])
-	buf = buf[4:]
-	if nullSize > 0 {
-		nullBuf := buf[:nullSize]
-		nulls := roaring64.New()
-		r := bytes.NewBuffer(nullBuf)
-		if _, err = nulls.ReadFrom(r); err != nil {
-			if n != nil {
-				vec.GetAllocator().Free(n)
+	// The downstream vector, ie CN vector needs isNull as argument.
+	// So, we can't directly call cn_vector.Append() without parsing the data.
+	// Hence, we are using src.Get(i) to retrieve the Null value as such from the src, and inserting
+	// it into the current CnVectorAdapter via this function.
+	for i := srcOff; i < srcOff+srcLen; i++ {
+		vec.Append(src.Get(i))
+	}
+}
+
+func (vec *vector[T]) forEachWindowWithBias(offset, length int, op ItOp, sels *roaring.Bitmap, bias int, shallow bool) (err error) {
+	//TODO: Benchmark.
+	if !vec.Nullable() {
+		var v T
+		if _, ok := any(v).([]byte); !ok {
+			// Optimization for :- Vectors which are 1. not nullable & 2. not byte[]
+			slice := vec.Slice().([]T)
+			slice = slice[offset+bias : offset+length+bias]
+			if sels == nil || sels.IsEmpty() {
+				for i, elem := range slice {
+					if err = op(elem, i+offset); err != nil {
+						break
+					}
+				}
+			} else {
+				idxes := sels.ToArray()
+				end := offset + length
+				for _, idx := range idxes {
+					if int(idx) < offset {
+						continue
+					} else if int(idx) >= end {
+						break
+					}
+					if err = op(slice[int(idx)-offset], int(idx)); err != nil {
+						break
+					}
+				}
 			}
 			return
 		}
-		vec.nulls = nulls
+
+	}
+	if sels == nil || sels.IsEmpty() {
+		for rowId := offset; rowId < offset+length; rowId++ {
+			var elem any
+			if shallow {
+				elem = vec.ShallowGet(rowId + bias)
+			} else {
+				elem = vec.Get(rowId + bias)
+			}
+			if err = op(elem, rowId); err != nil {
+				break
+			}
+		}
+	} else {
+
+		selsArray := sels.ToArray()
+		end := offset + length
+		for _, rowId := range selsArray {
+			if int(rowId) < offset {
+				continue
+			} else if int(rowId) >= end {
+				break
+			}
+			var elem any
+			if shallow {
+				elem = vec.ShallowGet(int(rowId) + bias)
+			} else {
+				elem = vec.Get(int(rowId) + bias)
+			}
+			if err = op(elem, int(rowId)); err != nil {
+				break
+			}
+		}
 	}
-	vec.roStorage = n
 	return
 }
 
-func (vec *vector[T]) Foreach(op ItOp, sels *roaring.Bitmap) (err error) {
-	return vec.impl.ForeachWindow(0, vec.Length(), op, sels)
-}
+// TODO: --- Need advise on the below functions
 
-func (vec *vector[T]) ForeachWindow(offset, length int, op ItOp, sels *roaring.Bitmap) (err error) {
-	return vec.impl.ForeachWindow(offset, length, op, sels)
+func (vec *vector[T]) Compact(deletes *roaring.Bitmap) {
```

Comment:
If there are 10,000 rows in the vector, only one-thousandth or one ten-thousandth of the deletion will cause performance problems in this way?